File: lua/chronopulse_encoder.lua
local json = require("dkjson")

local function loadCipher(path)
    local f = assert(io.open(path, "r"))
    local text = f:read("*a")
    f:close()
    return json.decode(text)
end

local function encodeChronoPulse(text, cipher)
    local alphabet = cipher.Alphabet
    local symbols = {}
    for k, v in pairs(cipher.Symbols) do
        symbols[v.Represents] = k
    end

    local result = {}
    for i = 1, #text do
        local c = text:sub(i,i)
        local upper = c:upper()
        if alphabet[upper] then
            table.insert(result, alphabet[upper].."●")
        elseif c == " " then
            table.insert(result, "○")
        elseif c:match("[,.;:]") then
            table.insert(result, "•")
        elseif symbols[c] then
            table.insert(result, symbols[c])
        else
            table.insert(result, "?")
        end
    end
    return table.concat(result)
end

local cipher = loadCipher("../chronopulse.json")
io.write("Enter text to encode: ")
local text = io.read()
print(encodeChronoPulse(text, cipher))

File: csharp/ChronoPulseEncoder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

class ChronoPulseEncoder
{
    static void Main()
    {
        var cipherJson = File.ReadAllText("../chronopulse.json");
        var cipher = JsonSerializer.Deserialize<Dictionary<string, object>>(cipherJson);
        var alphabet = JsonSerializer.Deserialize<Dictionary<string, string>>(
            ((JsonElement)cipher["Alphabet"]).GetRawText()
        );
        var symbols = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(
            ((JsonElement)cipher["Symbols"]).GetRawText()
        );

        var symbolMap = new Dictionary<string, string>();
        foreach (var kv in symbols)
            symbolMap[kv.Value.GetProperty("Represents").GetString()] = kv.Key;

        Console.Write("Enter text to encode: ");
        string text = Console.ReadLine();
        Console.WriteLine(Encode(text, alphabet, symbolMap));
    }

    static string Encode(string text, Dictionary<string, string> alphabet, Dictionary<string, string> symbols)
    {
        string result = "";
        foreach (char c in text)
        {
            string ch = c.ToString();
            if (alphabet.ContainsKey(ch.ToUpper()))
                result += alphabet[ch.ToUpper()] + "●";
            else if (ch == " ")
                result += "○";
            else if (",.;:".Contains(ch))
                result += "•";
            else if (symbols.ContainsKey(ch))
                result += symbols[ch];
            else
                result += "?";
        }
        return result;
    }
}
